storage/innobase/trx/trx0purge.cc

```diff
@@ -571,42 +493,26 @@ inline dberr_t purge_sys_t::iterator::free_history_rseg(trx_rseg_t &rseg) const
   goto loop;
 }
 
-/** Cleanse purge queue to remove the rseg that reside in undo-tablespace
-marked for truncate.
-@param[in]	space	undo tablespace being truncated */
-static void trx_purge_cleanse_purge_queue(const fil_space_t& space)
+void purge_sys_t::cleanse_purge_queue(const fil_space_t &space)
 {
-	typedef	std::vector<TrxUndoRsegs>	purge_elem_list_t;
-	purge_elem_list_t			purge_elem_list;
-
-	mysql_mutex_lock(&purge_sys.pq_mutex);
-
-	/* Remove rseg instances that are in the purge queue before we start
-	truncate of corresponding UNDO truncate. */
-	while (!purge_sys.purge_queue.empty()) {
-		purge_elem_list.push_back(purge_sys.purge_queue.top());
-		purge_sys.purge_queue.pop();
-	}
+  typedef std::vector<decltype(purge_queue)::value_type> purge_elem_list_t;
+  purge_elem_list_t purge_elem_list;
 
-	for (purge_elem_list_t::iterator it = purge_elem_list.begin();
-	     it != purge_elem_list.end();
-	     ++it) {
+  mysql_mutex_lock(&pq_mutex);
 
-		for (TrxUndoRsegs::iterator it2 = it->begin();
-		     it2 != it->end();
-		     ++it2) {
-			if ((*it2)->space == &space) {
-				it->erase(it2);
-				break;
-			}
-		}
+  /* Remove rseg instances that are in the purge queue before we start
+  truncate of corresponding UNDO truncate. */
+  while (!purge_queue.empty())
+  {
+    purge_elem_list.push_back(purge_sys.purge_queue.top());
+    purge_sys.purge_queue.pop();
+  }
```

Comment:
This logic can be simplified if we [implemented our own priority queue](https://www.geeksforgeeks.org/how-to-implement-priority-queue-using-heap-or-array/) on top of an array.

Note: b3c9cc6f212c650a18ca191a8889e2af68ffc9d6 or mysql/mysql-server@9d6d1902e091c868bb288e0ccf9f975ccb474db9 had added a "binary heap"; the header `ut0bh.h` that those commits refer to was apparently added later. As far as I understand, it is based on an array. The interesting part are the function `ib_bh_push()` and `ib_bh_pop()`. https://www.geeksforgeeks.org/priority-queue-using-binary-heap/ describes this technique.

This loop could access the underlying elements (indexes into `trx_sys.rseg_array`) directly, and then set the number of elements to 0. That would avoid restructuring the heap in the equivalent of `ib_bh_pop()`.

I realize that we do have to construct something like `purge_elem_list` and can’t easily modify `purge_queue` (this member function should not refer to `purge_sys`!) in place. However, the intermediate copy could be a simple `memcpy()` of the underlying array of up to 128 bytes. We could also filter for `&space` already during this first copying loop, and then just insert the surviving copied elements back to `purge_queue`.